### ISSUE TITLE
Handlers with parameters and static files

### ISSUE THREAD

**Issue by miyconst** [Starcounter/Starcounter#3691]
In my app I have a handler:

```cs
Handle.GET("/MyApp/{?}", (string key) => new MyPage() { Key = key });
```

Then I have some static files:

```
| wwwroot
| -- MyApp
| ---- viewmodels
| ------- MyPage.html
```

In the result there is no way to get the static file, because request like below hit the handler and stack in there. I would expect my handler be called only if there is no such static file.

```
/MyApp/viewmodels/MyPage.html
```

cc @alemoi

**Comment by alemoi** [comment]
Its how URI matcher works: the handler is determined on the gateway by longest matching path. File handler is just yet another handler. This is an architectural behavior and how URI matcher functions.
The same example:
```
/app1/handler1/{?}
/app1/{?}
```
you get request "/app1/handler1/something" which ends in `/app1/handler1/{?}` but you want to call "/app1/handler1/blabla" from the handler so it ends up in "/app1/{?}" handler. Currently its not possible, since longest matched URI handler will be called.

**Comment by miyconst** [comment]
@alemoi sorry, but I don't get you.

In this case:

```
GET: /app1/handler1/something
```

```
/app1/handler1/{?}
/app1/{?}
```

The longest matching handler would be `/app1/handler1/{?}`, why `/app1/{?}` is called?

**Comment by alemoi** [comment]
> The longest matching handler would be /app1/handler1/{?}, why /app1/{?} is called?

No, that's the thing that "/app1/{?}" is never called in current architecture.

**Comment by miyconst** [comment]
Okay, if the static files are also handlers, then

```
/MyApp/viewmodels/MyPage.html
```

Is much longer than:

```
/MyApp/{?}
```

Am I missing something?

**Comment by alemoi** [comment]
> Am I missing something?

No. In your example above, I suppose the static file handler is less generic than your other handler, so that's why the static file is never retriewed.

**Comment by miyconst** [comment]
Okay, so the only solution is to put my static files into a folder like `MyAppStatic/viewmodels/MyPage.html` or define more specific handlers. Right?

**Comment by alemoi** [comment]
For now I would say so: handlers should not overlap on a matching basis, with more specific being called.

**Comment by miyconst** [comment]
I am not sure this is something right to request, but I get used in asp.net that first are static files served. If there is no such file - check the handlers.

Still, there is a setting to change the behavior to the one we have.
